onnxruntime/core/graph/function.cc

```diff
@@ -7,6 +7,126 @@
 #include "core/graph/model.h"
 #include "onnx/shape_inference/implementation.h"
 
+namespace ONNX_NAMESPACE {
+// Infer shape for functions. This also supports 
+// nested model local functions.
+// TODO: Add this to onnx instead of adding it here.
+void InferShapeForFunctionNode(
+    InferenceContext& ctx,
+    const FunctionProto& func,
+    const std::unordered_map<std::string, int>& func_opset_imports,
+    const ShapeInferenceOptions& options,
+    const ISchemaRegistry* schema_registry,
+    const std::unordered_map<std::string, const FunctionProto*>& in_model_functions) {
+  GraphProto g;
+  // Get a temporary tensor-shape map
+  const auto num_func_inputs = func.input_size();
+  std::unordered_map<std::string, TypeProto*> valueTypesByName;
+  std::vector<TypeProto> types_cache(func.input_size());
+  for (int i = 0; i < num_func_inputs; ++i) {
+    types_cache[i] = *ctx.getInputType(i);
+    valueTypesByName[func.input().Get(i)] = &types_cache[i];
+  }
+
+  // Get a temporary initial value map
+  std::unordered_map<std::string, const TensorProto*> initializersByName;
+  std::unordered_map<std::string, const SparseTensorProto*> sparseInitializersByName;
+  for (int i = 0; i < static_cast<int>(ctx.getNumInputs()) && i < num_func_inputs; ++i) {
+    const TypeProto* type = ctx.getInputType(i);
+    if (type->value_case() == TypeProto::kTensorType && ctx.getInputData(i) != nullptr) {
+      initializersByName[func.input().Get(i)] = ctx.getInputData(i);
+    } else if (type->value_case() == TypeProto::kSparseTensorType &&
+               ctx.getInputSparseData(i) != nullptr) {
+      sparseInitializersByName[func.input().Get(i)] = ctx.getInputSparseData(i);
+    }
+  }
+  std::unordered_map<std::string, const AttributeProto*> attr_map;
+  for (auto& attr : func.attribute()) {
+    if (ctx.getAttribute(attr) != nullptr) {
+      attr_map[attr] = ctx.getAttribute(attr);
+    }
+  }
+
+  for (auto& n : func.node()) {
+    NodeProto copy_n(n);
+    // Add attribute information into the temporary node
+    copy_n.clear_attribute();
+    for (const auto& attr : n.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          auto copy_attr = *attr_map[attr.ref_attr_name()];
+          copy_attr.set_name(attr.name());
+          copy_n.add_attribute()->CopyFrom(copy_attr);
+        }
+      } else {
+        copy_n.add_attribute()->CopyFrom(attr);
+      }
+    }
+    ONNX_NAMESPACE::shape_inference::InferenceContextImpl func_node_ctx(
+        copy_n, valueTypesByName, initializersByName, sparseInitializersByName, {});
+
+    // Resolve domain for node
+    auto it = func_opset_imports.find(n.domain());
+    if (it == func_opset_imports.end()) {
+      return;
+    }
```

Comment:
do we get an error later so don't need one here?

Reply:
The type and shape inference wont be complete and the caller can decide whether to fail or proceed. If type and shape info is already available in the model there is no reason to fail therefore we dont want to fail here.

Reply:
Does the caller get a direct signal it failed? Or they have to infer that and the reason why it failed?

My concern is more with a silent exit from the inferencing. How would I debug an unexpected failure in that?

Reply:
In ORT we see a failure before we reach this point. ONNX_NAMESPACE::checker::[check_node](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/graph/graph.cc#L2396) for such a node without opset import for domain will fail. This is the error message I saw from my quick test: "Resolve subgraph failed:This is an invalid model. Error in Node:00000183A8C3B300 : No opset import for domain 'dom'"

In light of this I think it is OK to fail here as well...